keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2346 @@
+	// decide on which node a Pod should run, it will
+	// ask the driver(s) on which nodes the resource might be
+	// made available. To trigger that check, the scheduler
+	// provides the names of nodes which might be suitable
+	// for the Pod. Will be updated periodically until
+	// the claim is allocated.
+	//
+	// The ResourceClass.SuiteableNodes node selector can be
+	// used to filter out nodes based on labels. This prevents
+	// adding nodes here that the driver then would need to
+	// reject through UnsuitableNodes.
+	PotentialNodes []string
+
+	// Each resource driver is responsible for providing information about
+	// those claims in the Pod that the driver manages. It can skip
+	// adding that information when it already allocated the claim.
+	//
+	// +listType=map
+	// +listMapKey=podResourceClaimName
+	// +optional
+	Claims []ResourceClaimScheduling
+}
+
+// ResourceClaimScheduling contains information about one
+// particular claim in a pod while scheduling that pod.
+type ResourceClaimScheduling struct {
+	// PodResourceClaimName matches the PodResourceClaim.Name field.
+	PodResourceClaimName string
+
+	// A change of the PotentialNodes field in the PodScheduling object
+	// triggers a check in the driver
+	// on which of those nodes the resource might be made available. It
+	// then excludes nodes by listing those where that is not the case in
+	// UnsuitableNodes.
+	//
+	// Nodes listed here will be ignored by the scheduler when selecting a
+	// node for a Pod. All other nodes are potential candidates, either
+	// because no information is available yet or because allocation might
+	// succeed.
+	//
+	// This can change, so the driver must refresh this information
+	// periodically and/or after changing resource allocation for some
+	// other ResourceClaim until a node gets selected by the scheduler.
+	UnsuitableNodes []string
+}
+
+type PodSpec {
+   ...
+   // ResourceClaims defines which ResourceClaims must be allocated
+   // and reserved before the Pod is allowed to start. The resources
+   // will be made available to those containers which reference them
+   // by name.
+   //
+   // At most 32 entries are allowed.
+   ResourceClaims []PodResourceClaim
+   ...
+}
+
+type  ResourceRequirements {
+   Limits ResourceList
+   Requests ResourceList
+   ...
+   // The entries are the names of resources in PodSpec.ResourceClaims
+   // that are used by the container.
+   Claims []string
+   ...
+}
+
+// PodResourceClaim references exactly one ResourceClaim, either by name or
+// by embedding a template for a ResourceClaim that will get created
+// by the resource claim controller in kube-controller-manager.
+type PodResourceClaim struct {
+	// A name under which this resource can be referenced by the containers.
+	Name string
+
+	// The resource is independent of the Pod and defined by
+	// a separate ResourceClaim in the same namespace as
+	// the Pod. Either this or Template must be set, but not both.
+	ResourceClaimName *string
```

Comment:
We need to think carefully about one-of semantics here.  What if neither are specified - how should consumers react?  Will we EVER need to add another option here?  History shows that when we say "no" to this question, we are often wrong :)

Reply:
This would be the same as with volume type: when a new alternative gets added, old code won't see it and must treat an "empty" struct as "I don't know how to deal with this" and error out.

Reply:
ACK - so we should; a) document it as such; b) insulate against grwoing extra fields here.  e.g.

```
{
    name: foobar
    claim: {
        // one of these
        resourceClaimName: abc123
        template: { ... }
    }
}

Reply:
I added `ClaimSource` (matching `VolumeSource) with just the two fields and additional comments about future extensions:

```
    // Claim determines where to find the claim.
    Claim ClaimSource
}

// ClaimSource either references one separate ResourceClaim by name or
// embeds a template for a ResourceClaim, but never both.
//
// Additional options might get added in the future, so code using this
// struct must error out when none of the options that it supports are set.
ClaimSource struct {
	// The resource is independent of the Pod and defined by
	// a separate ResourceClaim in the same namespace as
	// the Pod. Either this or Template must be set, but not both.
	ResourceClaimName *string

	// Will be used to create a stand-alone ResourceClaim to allocate the resource.
	// The pod in which this PodResource is embedded will be the
	// owner of the ResourceClaim, i.e. the ResourceClaim will be deleted together with the
	// pod.  The name of the ResourceClaim will be `<pod name>-<resource name>` where
	// `<resource name>` is the name PodResource.Name
	// Pod validation will reject the pod if the concatenated name
	// is not valid for a ResourceClaim (for example, too long).
	//
	// An existing ResourceClaim with that name that is not owned by the pod
	// will *not* be used for the pod to avoid using an unrelated
	// resource by mistake. Scheduling is then blocked until
	// the unrelated ResourceClaim is removed. If such a pre-created ResourceClaim is
	// meant to be used by the pod, the ResourceClaim has to be updated with an
	// owner reference to the pod once the pod exists. Normally
	// this should not be necessary, but it may be useful when
	// manually reconstructing a broken cluster.
	//
	// Running the pod also gets blocked by a wrong ownership. This should
	// be even less likely because of the prior scheduling check, but could
	// happen if a user force-deletes or modifies the ResourceClaim.
	//
	// This field is read-only and no changes will be made by Kubernetes
	// to the ResourceClaim after it has been created.
	// Either this or ResourceClaimName must be set, but not both.
	Template *ResourceClaimTemplate
}
```